### PR TITLE
Queue Implementation

### LINKED ISSUE #126: [FEAT]: Implement Queue Data Structure
## 📘 Description

We need to implement a Queue data structure that enables efficient FIFO (First In First Out) operations. A Queue is a fundamental data structure that maintains elements in a specific order where the first element added is the first one to be removed. This implementation will be part of our data structures package and should be optimized for performance while maintaining code readability.

## ✅ Acceptance Criteria

Implement core operations:

- enqueue: add an element to the back of the queue
- dequeue: remove and return the front element of the queue
- front: peek at the front element without removing it
- is_empty: check if the queue is empty
- size: get the current number of elements

Include usage examples and documentation
Add unit tests for all operations

## 🌎 References

- [Alexandria Queue Data Structure](https://github.com/keep-starknet-strange/alexandria/blob/main/packages/data_structures/src/queue.cairo)

## 📜 Additional Notes

- ⚠ **Do not apply until the ODBoost begins.**
- ⚠ **Read our contribution guidelines before applying.**

### PR DESCRIPTION
**Pull request by armaanansari121**
# 📝 Queue Implementation 

## 🛠️ Issue
- Closes #126 

## 📖 Description
Implemented the required functions to efficiently store and manage a queue data structure. 
Also added the necessary tests for each function and edge cases.

## 🖼️ Screenshots (if applicable)
image

### CODE REVIEW

File: cairo/scripts/queue_data_structure/src/queue.cairo [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    pub data: Felt252Dict<Nullable<T>>,
+    /// Index of the front element
+    pub front: usize,
+    /// Index where the next element will be inserted
+    pub back: usize,
+}
+
+pub impl DestructQueue<T, +Drop<T>> of Destruct<Queue<T>> {
+    fn destruct(self: Queue<T>) nopanic {
+        self.data.squash();
```
Comment (danielcdz) [inline cairo/scripts/queue_data_structure/src/queue.cairo:51]: What .squash() does here?

File: cairo/scripts/queue_data_structure/tests/test_queue.cairo [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        assert_eq!(queue.dequeue(), Option::Some(i));
+    };
+}
+
+#[test]
+fn test_destruct() {
+    let mut queue: Queue<u32> = QueueImpl::new();
+    queue.enqueue(1);
+    queue.enqueue(2);
+    queue.destruct();
```
Comment (danielcdz) [inline cairo/scripts/queue_data_structure/tests/test_queue.cairo:71]: Is there any assertion we can do here to confirm that the queue was destroyed?

**Comment by armaanansari121** [comment]
@coxmars This is what I found in the cairo book. 
image
So what I understand from here is that doing all of that manually was unnecessary since Cairo takes care of it if the Destruct Trait is implemented. 
So I think I should update my code to have Destruct trait implemented on the Queue struct. 
Let me know if you want that or want it to be how it is currently. 
I personally think there's no need to handle things manually when Cairo does it the best for us.

File: cairo/scripts/queue_data_structure/tests/test_queue.cairo (hunk 2 above)
Comment (armaanansari121) [reply cairo/scripts/queue_data_structure/tests/test_queue.cairo:71]: I did try to make some assertions, but the variable is moved after this.

File: cairo/scripts/queue_data_structure/src/queue.cairo (hunk 1 above)
Comment (armaanansari121) [reply cairo/scripts/queue_data_structure/src/queue.cairo:51]: I think it is supposed to verify the dictionary entries and lower the size of the dictionary. 
image

Comment (danielcdz) [reply cairo/scripts/queue_data_structure/src/queue.cairo:51]: Hey thanks for attaching this

**Comment by danielcdz** [comment]
@armaanansari121 you are great, thanks for the explanation, we can leave this as it is, thanks for your awesome job!

**Review by danielcdz — APPROVED** [review]

**Comment by armaanansari121** [comment]
Thank you very much. Loved contributing to this project. 
I hope to be back for more. Hopefully some DeFi related stuff 👀
